mmengine/optim/scheduler/param_scheduler.py

```diff
@@ -1283,3 +1283,275 @@ def get_position_from_periods(
+        self.cooldown = cooldown
+        self.cooldown_counter = 0
+        self.rule_worse = None  # the worse value for the chosen mode
+        self.best = None
+        self.num_bad_epochs = 0
+        self.eps = eps
+
+        self.monitor = monitor
+        self._init_is_better(
+            rule=rule, threshold=threshold, threshold_rule=threshold_rule)
+        self._reset()
+
+        # remove call self.step() and init self._global_step = 0
+        self._last_value = [
+            group[self.param_name] for group in self.optimizer.param_groups
+        ]
+
+    def step(self, metrics=None):
+        """Adjusts the parameter value of each parameter group based on the
+        specified schedule.
+
+        Args:
+            metrics (Dict[str, float], optional): Evaluation results of all
+                metrics on validation dataset. The keys are the names of the
+                metrics, and the values are corresponding results.
+                Defaults to None.
+        """
+        if metrics is None:
+            # only to count self._global_step
+            self._global_step += 1
+            return
+
+        if not isinstance(metrics, dict):
+            raise TypeError('metrics type should be dict,'
+                            f' but got type {type(metrics)}')
+
+        # Compute parameter value per param group in the effective range
+        if self.begin <= self._global_step < self.end:
+            self.last_step += 1
+
+            # convert `metric` to float, in case it's a zero-dim Tensor
+            metric = metrics.get(self.monitor, None)
+            if metric is not None:
+                if self._is_better(metric, self.best):
+                    self.best = metric
+                    self.num_bad_epochs = 0
+                else:
+                    self.num_bad_epochs += 1
+
+                if self._in_cooldown():
+                    self.cooldown_counter -= 1
+                    self.num_bad_epochs = 0  # ignore bad epochs in cooldown
+
+                if self.num_bad_epochs > self.patience:
+                    values = self._get_value()
+
+                    for i, data in enumerate(
+                            zip(self.optimizer.param_groups, values)):
+                        param_group, value = data
+                        if param_group[self.param_name] - value > self.eps:
+                            param_group[self.param_name] = value
+                            self.print_value(self.verbose, i, value)
+                    self.cooldown_counter = self.cooldown
+                    self.num_bad_epochs = 0
+
+            else:
+                raise KeyError(f'Excepted key in {list(metrics.keys())},'
+                               f' but got key {self.monitor} is not in dict')
+
+        self._last_value = [
+            group[self.param_name] for group in self.optimizer.param_groups
+        ]
+
+    def print_value(self, is_verbose: bool, group: int, value: float):
```

Comment:
```suggestion
    def print_value(self, is_verbose: bool, group: int, value: float) -> None:
```

Reply:
How about directly refining the `_ParameterScheduler.print_value` rather than overriding it?

Reply:
Because the scheduler won't print every epoch, we should show the step which reduces parameters. And `_ParameterScheduler.print_value` don't show the step.